pkg/apis/kubeonecluster/types.go

```diff
@@ -0,0 +1,181 @@
+type APIEndpoint struct {
+	// Host is the hostname on which API is running
+	Host string `json:"host"`
+
+	// Port is the port used to reach to the API
+	Port string `json:"port"`
+}
+
+// ProviderName represents the name of a provider
+type ProviderName string
+
+// ProviderName values
+const (
+	ProviderNameAWS          ProviderName = "aws"
+	ProviderNameOpenStack    ProviderName = "openstack"
+	ProviderNameHetzner      ProviderName = "hetzner"
+	ProviderNameDigitalOcean ProviderName = "digitalocean"
+	ProviderNameVSphere      ProviderName = "vsphere"
+	ProviderNameGCE          ProviderName = "gce"
+	ProviderNameNone         ProviderName = "none"
+)
+
+// ProviderConfig describes the provider that is running the machines
+type ProviderConfig struct {
+	Name        ProviderName `json:"name"`
+	External    bool         `json:"external"`
+	CloudConfig string       `json:"cloudConfig"`
+}
+
+// VersionConfig describes the versions of components that are installed on the machines
+type VersionConfig struct {
+	// TODO(xmudrii): switch to semver
+	Kubernetes string `json:"kubernetes"`
+}
+
+// ClusterNetworkConfig describes the cluster network
+type ClusterNetworkConfig struct {
+	PodSubnet         string `json:"podSubnet"`
+	ServiceSubnet     string `json:"serviceSubnet"`
+	ServiceDomainName string `json:"serviceDomainName"`
+	NodePortRange     string `json:"nodePortRange"`
+}
+
+// ProxyConfig configures proxy for the Docker daemon and is used by KubeOne scripts
+type ProxyConfig struct {
+	HTTP    string `json:"http"`
+	HTTPS   string `json:"https"`
+	NoProxy string `json:"noProxy"`
+}
+
+// WorkerConfig describes a set of worker machines
+type WorkerConfig struct {
+	Name     string       `json:"name"`
+	Replicas *int         `json:"replicas"`
+	Config   ProviderSpec `json:"providerSpec"`
+}
+
+// ProviderSpec describes a worker node
+type ProviderSpec struct {
+	CloudProviderSpec   *runtime.RawExtension `json:"cloudProviderSpec"`
+	Labels              map[string]string     `json:"labels"`
+	SSHPublicKeys       []string              `json:"sshPublicKeys"`
+	OperatingSystem     string                `json:"operatingSystem"`
+	OperatingSystemSpec *runtime.RawExtension `json:"operatingSystemSpec"`
+}
+
+// MachineControllerConfig configures kubermatic machine-controller deployment
+type MachineControllerConfig struct {
+	Deploy *bool `json:"deploy"`
+	// Provider is provider to be used for machine-controller
+	// Defaults and must be same as chosen cloud provider, unless cloud provider is set to None
+	Provider    ProviderName      `json:"provider"`
+	Credentials map[string]string `json:"credentials"`
+}
+
+// Features controls what features will be enabled on the cluster
+type Features struct {
+	PodSecurityPolicy *PodSecurityPolicy `json:"podSecurityPolicy"`
+	DynamicAuditLog   *DynamicAuditLog   `json:"dynamicAuditLog"`
+	MetricsServer     *MetricsServer     `json:"metricsServer"`
+	OpenIDConnect     *OpenIDConnect     `json:"openidConnect"`
```

Comment:
What's up with all those pointers?

Please don't forget to update type definitions from [pkg/config/cluster.go](https://github.com/kubermatic/kubeone/blob/f5067336c5b7afaaaf4072a8a44ceb35207cc5aa/pkg/config/cluster.go#L342-L381)